cognite/client/_api_client.py

```diff
@@ -797,6 +807,98 @@ def _update_multiple(
+            ]
+
+            created: T_CogniteResourceList | None = None
+            updated: T_CogniteResourceList | None = None
+            try:
+                if to_create:
+                    created = self._create_multiple(
+                        to_create, list_cls=list_cls, resource_cls=resource_cls, input_resource_cls=input_resource_cls
+                    )
+                if to_update:
+                    updated = self._update_multiple(
+                        to_update,
+                        list_cls=list_cls,
+                        resource_cls=resource_cls,
+                        update_cls=update_cls,
+                        mode=mode,
+                    )
+            except CogniteAPIError as api_error:
+                successful = list(api_error.successful)
+                unknown = list(api_error.unknown)
+                failed = list(api_error.failed)
+
+                successful.extend(not_found_error.successful)
+                unknown.extend(not_found_error.unknown)
+                if created is not None:
+                    # The update call failed
+                    successful.extend([item.external_id for item in created])
+                if updated is None and created is not None:
+                    # The created call failed
+                    failed.extend([item.external_id if item.external_id is not None else item.id for item in to_update])
+                raise CogniteAPIError(
+                    api_error.message, code=api_error.code, successful=successful, failed=failed, unknown=unknown
+                )
+            # Need to retrieve the successful updated items from the first call.
+            successful_resources: T_CogniteResourceList | None = None
+            if not_found_error.successful:
+                identifiers = IdentifierSequence.of(*not_found_error.successful)
+                successful_resources = self._retrieve_multiple(
+                    list_cls=list_cls, resource_cls=resource_cls, identifiers=identifiers
+                )
+                if isinstance(successful_resources, resource_cls):
+                    successful_resources = list_cls([successful_resources], cognite_client=self._cognite_client)
+
+            result = list_cls(
+                (successful_resources or []) + (created or []) + (updated or []), cognite_client=self._cognite_client
+            )
+
+            # Reorder to match the order of the input items
+            result_by_identifier = {
+                identifier: item
+                for item in result
+                for identifier in [item.external_id, item.id]
+                if identifier is not None
+            }
+            result = list_cls(
+                [result_by_identifier[item.external_id if item.external_id is not None else item.id] for item in items],
+                cognite_client=self._cognite_client,
+            )
```

Comment:
The `result` variable already has the mapping stored on the object (to power the `.get` method), so you just need the reorder-code:

Here's a suggestion 🤓 
```suggestion
            # Reorder to match the order of the input items
            result.data = [
                result.get(**Identifier.load(item.id, item.external_id).as_dict(camel_case=False)) for item in items
            ]
```